### PR TITLE
(PUP-11683) Ignore invalid tasks when listing

### PR DESCRIPTION
**Pull request by justinstoller**
Previously, if a task failed validation in someway, typically with bad metadata, then `Puppet::InfoService.tasks_per_environment` method would error, failing to return any task info.

The desired behavior is to still list valid tasks in an environment, even if one task is invalid.

### CODE REVIEW

**Comment by justinstoller** [comment]
Actually I think 7.x is the desired branch, right @donoghuc ?

File: lib/puppet/info_service/task_information_service.rb [hunk 1]
```diff
@@ -4,11 +4,17 @@ class Puppet::InfoService::TaskInformationService
...
+
     env.modules.map do |mod|
       mod.tasks.map do |task|
-        {:module => {:name => task.module.name}, :name => task.name, :metadata => task.metadata}
+        # If any task is malformed continue to list other tasks in module
+        begin
+          task.validate
+          {:module => {:name => task.module.name}, :name => task.name, :metadata => task.metadata}
+        rescue Puppet::Module::Task::Error => _
+        end
```
Comment (joshcooper) [inline lib/puppet/info_service/task_information_service.rb:15]: Can we log a warning with the error and have an implicit `nil` at the end so it's clear it's compacted later.

**Comment by justinstoller** [comment]
Not sure what's up with the older JRuby failure, will investigate if it repeats.

**Comment by justinstoller** [comment]
I had to rerun the tests again, this time Windows failed with unrelated changes.

**Review by joshcooper — APPROVED** [review]

File: spec/unit/info_service_spec.rb [hunk 2]
```diff
@@ -33,6 +33,26 @@
...
+                                                                            {:name => 'btask.json',
+                                                                             :content => metadata.to_json}],
+                                                                           ['ctask',
+                                                                            {:name => 'ctask.json',
+                                                                             :content => metadata.to_json}]]})
+          File.write("#{modpath}/#{mod_name}/tasks/atask.json", "NOT JSON")
+          expect(Puppet).to receive(:send_log).with(:err, 'Failed to validate task')
+          @tasks = Puppet::InfoService.tasks_per_environment(env_name)
+
+          expect(@tasks.map{|t| t[:name]}.sort).to eq(%w{test1::btask test1::ctask})
```
Comment (joshcooper) [inline spec/unit/info_service_spec.rb:52]: not blocking, but `expect([1, 2]).to contain_exactly(2, 1)` can be used to match arrays ignoring order

**Pull request merged**